### PR TITLE
Use network status flow to know when we have internet or not

### PR DESCRIPTION
**Pull request by tevincent**
depends on https://github.com/Infomaniak/android-core/pull/208

https://github.com/user-attachments/assets/0c0d365b-2b22-4dcf-b5d9-652c2ab3cc51

### CODE REVIEW

File: app/src/main/java/com/infomaniak/drive/ui/MainViewModel.kt [hunk 1]
```diff
@@ -557,6 +561,24 @@ class MainViewModel(
...
+                    level = if (isAvailable) SentryLevel.INFO else SentryLevel.WARNING
+                })
+                isInternetAvailable.postValue(isAvailable)
+                if (isAvailable) {
+                    AccountUtils.updateCurrentUserAndDrives(this@MainViewModel.getContext())
+                    restartUploadWorkerIfNeeded()
+                }
+            }
+        }
+    }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/MainViewModel.kt:580]:
```suggestion
    fun observeNetworkStatus() = viewModelScope.launch(Dispatchers.IO) {
        getContext().networkStatusFlow().collect { isAvailable ->
            SentryLog.d("Internet availability", if (isAvailable) "Available" else "Unavailable")
            Sentry.addBreadcrumb(
                Breadcrumb().apply {
                    category = "Network"
                    message = "Internet access is available : $isAvailable"
                    level = if (isAvailable) SentryLevel.INFO else SentryLevel.WARNING
                },
            )
            isInternetAvailable.postValue(isAvailable)
            if (isAvailable) {
                AccountUtils.updateCurrentUserAndDrives(this@MainViewModel.getContext())
                restartUploadWorkerIfNeeded()
            }
        }
    }
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Review by sirambd — CHANGES_REQUESTED** [review]
At the moment, `mainViewModel.isNetworkAvailable.value` doesn't seem to work very well

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]
The previous solution was fine.

Now, we are doing :
``` kotlin
viewModelScope.launch {
    isNetworkAvailable.collect {
        …
    }
}
```
everywhere in the UI.

That's pretty ugly.
Why this solution ?

**Review by KevinBoulongne — APPROVED** [review]

File: app/src/main/java/com/infomaniak/drive/views/FileInfoActionsView.kt [hunk 2]
```diff
@@ -99,25 +100,29 @@ class FileInfoActionsView @JvmOverloads constructor(
 
     // TODO - Enhanceable code : Replace these let by an autonomous view with "enabled/disabled" method ?
     private fun computeFileRights(file: File, rights: Rights) = with(binding) {
-        val isOnline = mainViewModel.isInternetAvailable.value == true
-
-        displayInfo.isEnabled = isOnline
-        disabledInfo.isGone = isOnline
+        ownerFragment.lifecycleScope.launch {
+            mainViewModel.isNetworkAvailable.collect { isNetworkAvailable ->
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/views/FileInfoActionsView.kt:104]: Use `.value` instead of collect ?

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/drive/ui/home/HomeFragment.kt [hunk 3]
```diff
@@ -52,9 +54,12 @@ class HomeFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) = with(binding) {
         super.onViewCreated(view, savedInstanceState)
 
-        mainViewModel.isInternetAvailable.observe(viewLifecycleOwner) { isInternetAvailable ->
-            noNetworkCard.root.isGone = isInternetAvailable
+        lifecycleScope.launch {
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/ui/home/HomeFragment.kt:57]: use `viewLifeCycleOwner.lifecyleScope` instead

File: app/src/main/java/com/infomaniak/drive/ui/fileList/FileListFragment.kt [hunk 4]
```diff
@@ -832,26 +836,28 @@ open class FileListFragment : MultiSelectFragment(MATOMO_CATEGORY), SwipeRefresh
         changeControlsVisibility: Boolean,
         ignoreOffline: Boolean = false
     ) {
-        if (_binding == null) return
+        lifecycleScope.launch {
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/FileListFragment.kt:839]: use `viewLifecycleOwner.lifecycleScope` instead

File: app/src/main/java/com/infomaniak/drive/ui/fileList/FileListFragment.kt [hunk 5]
```diff
@@ -670,16 +671,19 @@ open class FileListFragment : MultiSelectFragment(MATOMO_CATEGORY), SwipeRefresh
...
-
-        fileListViewModel.getFiles(
-            folderId,
-            order = fileListViewModel.sortType,
-            sourceRestrictionType = if (isNetworkUnavailable) ONLY_FROM_LOCAL else sourceRestrictionType,
-            userDrive = userDrive,
-            isNewSort = isNewSort,
-        ).observe(viewLifecycleOwner) {
-            onFinish?.invoke(it)
+        lifecycleScope.launch {
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/FileListFragment.kt:674]: This `launch` is really useful ?

File: app/src/main/java/com/infomaniak/drive/ui/fileList/FileListFragment.kt [hunk 6]
```diff
@@ -75,6 +75,7 @@ import com.infomaniak.drive.views.NoItemsLayoutView
 import com.infomaniak.lib.core.utils.*
 import com.infomaniak.lib.core.utils.Utils.createRefreshTimer
 import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.first
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/FileListFragment.kt:78]: I can't see where it's used

**Review by sirambd — CHANGES_REQUESTED** [review]
Do a global refactor to remove all useless imports

**Review by sirambd — APPROVED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Pull request merged**